lib/graphql/static_validation/rules/fields_will_merge.rb

```diff
@@ -2,48 +2,342 @@
+        find_conflicts_between(fields, fields2, mutually_exclusive: mutually_exclusive)
+
+        # (I) Then collect conflicts between the first collection of fields and
+        # those referenced by each fragment name associated with the second.
+        fragment_spreads2.each do |fragment_spread|
+          fragment_is_mutually_exclusive = mutually_exclusive?(
+            parent_type1,
+            fragment_spread.parent_type
+          )
+
+          find_conflicts_between_fields_and_fragment(
+            fragment_spread,
+            fields,
+            mutually_exclusive: mutually_exclusive || fragment_is_mutually_exclusive,
+          )
+        end
+
+        # (I) Then collect conflicts between the second collection of fields and
+        # those referenced by each fragment name associated with the first.
+        fragment_spreads.each do |fragment_spread|
+          fragment_is_mutually_exclusive = mutually_exclusive?(
+            parent_type2,
+            fragment_spread.parent_type
+          )
+
+          find_conflicts_between_fields_and_fragment(
+            fragment_spread,
+            fields2,
+            mutually_exclusive: mutually_exclusive || fragment_is_mutually_exclusive,
+          )
+        end
+
+        # (J) Also collect conflicts between any fragment names by the first and
+        # fragment names by the second. This compares each item in the first set of
+        # names to each item in the second set of names.
+        fragment_spreads.each do |frag1|
+          fragment_spreads2.each do |frag2|
+            fragments_are_mutually_exclusive = mutually_exclusive?(
+              frag1.parent_type,
+              frag2.parent_type
+            )
+
+            find_conflicts_between_fragments(
+              frag1,
+              frag2,
+              mutually_exclusive: mutually_exclusive || fragments_are_mutually_exclusive,
+            )
+          end
+        end
+      end
+
+      def find_conflicts_between(response_keys, response_keys2, mutually_exclusive:)
+        response_keys.each do |key, fields|
+          fields2 = response_keys2[key]
+          if fields2
+            fields.each do |field|
+              fields2.each do |field2|
+                find_conflict(
+                  key,
+                  field,
+                  field2,
+                  mutually_exclusive: mutually_exclusive,
+                )
               end
             end
-            args.uniq!
+          end
+        end
+      end
 
-            if args.length > 1
-              msg = "Field '#{node.name}' has an argument conflict: #{args.map{ |arg| GraphQL::Language.serialize(arg) }.join(" or ")}?"
-              context.errors << GraphQL::StaticValidation::Message.new(msg, nodes: node.ast_nodes.to_a)
-            end
+      def fields_and_fragments_from_selection(node, parent_type:)
+        @fields_and_fragments_from_node[node] ||= begin
```

Comment:
Fields and Fragment Spreads are collected once per node, iterating over the list only one.